gossip/bloom.go

```diff
@@ -0,0 +1,57 @@
+// Copyright (C) 2019-2023, Ava Labs, Inc. All rights reserved.
+// See the file LICENSE for licensing terms.
+
+package gossip
+
+import (
+	"encoding/binary"
+	"hash"
+
+	"github.com/ava-labs/avalanchego/ids"
+	bloomfilter "github.com/holiman/bloomfilter/v2"
+)
+
+const (
+	DefaultBloomM = 8 * 1024 // 8 KiB
+	DefaultBloomK = 4
+	// DefaultBloomMaxFilledRatio is the max ratio of filled slots in the bloom
+	// filter before we reset it to avoid too many collisions.
+	DefaultBloomMaxFilledRatio = 0.75
+)
+
+var _ hash.Hash64 = (*hasher)(nil)
+
+// ResetBloomFilterIfNeeded resets a bloom filter if it breaches a ratio of
+// filled elements. Returns true if the bloom filter was reset.
+func ResetBloomFilterIfNeeded(
+	bloomFilter **bloomfilter.Filter,
```

Comment:
Do we need a pointer to a pointer here?

Reply:
I wonder if we should just have a  `NeedReset(...) bool` function and recreate the bloom filter anew in the called?

Reply:
Yeah so this is a funny situation, we either pass a double pointer here, or we pass a single pointer here but this causes an issue since when we overwrite this pointer's value like so:
```
	fresh, _ := bloomfilter.New((*bloomFilter).M(), (*bloomFilter).K())
	*bloomFilter = *fresh // copying a mutex here
```

it flags the linter as us copying a mutex (I think this is technically safe since it's gonna be unlocked and no one else can modify it). I felt like the double pointer was less evil than this.

Reply:
> I wonder if we should just have a `NeedReset(...) bool` function and recreate the bloom filter anew in the called?

Yeah an alternative would be just to remove this function entirely, it's only used in two places and it's a small snippet of code so the duplication cost is really not bad.

Reply:
Ya I'd prefer to either remove this entirely or update this to a function that does not deal with a double pointer. Even though this seems correct, it still seems overly complicated. I think another simpler alternative would be to have a function that returns nil if no reset is needed and returns a non-nil pointer to a new bloom filter if it is needed (or a boolean if that seems cleaner)